doc/SAIP4Extensions/SAI_P4EXT_API_Proposal.md

```diff
@@ -0,0 +1,279 @@
+SAI P4EXT API Proposal
+=====================
+
+Title       | SAI IPsec API Proposal
+------------|----------------------
+Authors     | Intel.
+Status      | Draft
+Type        | Standards track
+Created     | 03/18/2021
+SAI-Version | ?
+-------------------------------
+
+# Scope #
+
+This document defines the technical specifications for the API used to support P4 Extensions in Open Compute Project Switch Abstraction Interface (SAI). P4 Extensions can be seen as an abstraction over niche/device specific capabilities. It facilitates programming of these new features via the familiar SAI API model using a single P4 Extension Object. The exact feature enabled via the P4 Extensions itself is outside the scope of this document and SAI P4 Extensions API.
+
+# Overview #
+
+SAI P4 extension introduces a single new SAI object SAI_OBJECT_TYPE_P4EXT_ENTRY of object_type_oid. This object along with its attributes provides an abstraction of target specific configurable features otherwise not described in SAI via the P4 match action paradigm in a feature agnostic manner. In contrast to other SAI objects which are mapped 1:1 to an entity in the SAI Pipeline Netwokring Object Model, for eg. VLAN, Neighbor, Mirror etc.; SAI P4 Extensions introduces 1:N mapping between the  SAI_OBJECT_TYPE_P4EXT_ENTRY object and niche features Feature1, Feature2, etc. This new paradigm enables the
+1. Ability to add exclusive features
+2. Expose device specific capabilities
+3. Rapid application prototyping
+4. Ability to introduce new features potentially without Recompiling SAI
+
+## Object Dependencies ##
+
+In the current SAI P4 Extension Proposal the newly introduced SAI_OBJECT_TYPE_P4EXT_ENTRY shall remain disconnected from the SAI Pipeline Object Graph. In other words there shall be no dependency from/to other SAI object to/from SAI_OBJECT_TYPE_P4EXT_ENTRY. Also the proposal places no restriction on the relative placement of the new features/table in the SAI Pipeline. The implementation and interaction of the new feature with the SAI pipeline is left to each individual vendor.Vendor A and Vendor B may choose to implement different sets of features using the new SAI P4 extension object. While a Vendor C may decide to not implement any new features using SAI P4 extensions.
```

Comment:
I think we should move this to a restrictions section later. No point confusing reader at this point.

Reply:
Actually I think I saw some proposals that explain the dependency of new object on other sai objects or vice versa upfront. So I decided to keep it at the top. This also sets the expectations correctly upfront, that the proposal does not allow for any cross references between P4 Extension object and other SAI objects

Reply:
Got it.

Reply:
"Vendor A and Vendor B may choose to implement different sets of features using the new SAI P4 extension object. While a Vendor C may decide to not implement any new features using SAI P4 extensions."
This whole section I think is unnecessary.